From 16cebd88bfdc005bc8fb07862e49cdedd5547d53 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:06 -0400
Subject: [usb] xhci: fix derivation of TRB's DMA address in xhci_log_event Trace Event Class

Message-id: <1410887463-20674-175-git-send-email-dzickus@redhat.com>
Patchwork-id: 95240
O-Subject: [RHEL7 PATCH 174/591] xhci: fix derivation of TRB's DMA address in xhci_log_event Trace Event Class
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit f00466963817c8a240f022af864cdc39d482cdb7
Author: Xenia Ragiadakou <burzalodowa@gmail.com>
Date:   Mon Sep 9 21:03:11 2013 +0300

    xhci: fix derivation of TRB's DMA address in xhci_log_event Trace Event Class

    This patch fixes the retrieval of the DMA address of the TRB that generated
    the event by converting the field[0] (low address bits) and field[1] (high
    address bits) to CPU byteorder and then typecasting field[1] to u64 so that
    the bitshift will not lead to overflow.
    In the original code, the typecasting of le32 to u64 was incorrect and the
    subsequent conversion to le64 reverts the low and high address parts.
    This bug was found using sparse.

    Signed-off-by: Xenia Ragiadakou <burzalodowa@gmail.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-trace.h b/drivers/usb/host/xhci-trace.h
index 20364cc..afe0933 100644
--- a/drivers/usb/host/xhci-trace.h
+++ b/drivers/usb/host/xhci-trace.h
@@ -120,8 +120,8 @@ DECLARE_EVENT_CLASS(xhci_log_event,
  ),
  TP_fast_assign(
   __entry->va = trb_va;
-  __entry->dma = le64_to_cpu(((u64)ev->field[1]) << 32 |
-      ev->field[0]);
+  __entry->dma = ((u64)le32_to_cpu(ev->field[1])) << 32 |
+     le32_to_cpu(ev->field[0]);
   __entry->status = le32_to_cpu(ev->field[2]);
   __entry->flags = le32_to_cpu(ev->field[3]);
   memcpy(__get_dynamic_array(trb), trb_va,
-- 
1.7.1